File: checkin-service/models/checkin.go
package models

import (
	"errors"
	"time"
)

// CheckIn statuses
const (
	CheckInStatusSuccess    = "success"
	CheckInStatusInvalid    = "invalid"
	CheckInStatusAlreadyUsed = "already_used"
	CheckInStatusCancelled  = "cancelled"
)

type CheckIn struct {
	ID          string     `db:"id"`
	TicketID    string     `db:"ticket_id"`
	EventID     string     `db:"event_id"`
	UserID      string     `db:"user_id"`
	StaffID     *string    `db:"staff_id"`
	QRCode      string     `db:"qr_code"`
	Status      string     `db:"status"`
	CheckInTime time.Time  `db:"check_in_time"`
	DeviceID    *string    `db:"device_id"`
	Gate        *string    `db:"gate"`
	Notes       *string    `db:"notes"`
	CreatedAt   time.Time  `db:"created_at"`
}

func (c *CheckIn) Validate() error {
	if c.TicketID == "" {
		return errors.New("ticket_id is required")
	}
	if c.EventID == "" {
		return errors.New("event_id is required")
	}
	if c.QRCode == "" {
		return errors.New("qr_code is required")
	}
	return nil
}

// EventStats holds aggregated check-in statistics for an event
type EventStats struct {
	EventID        string
	TotalCheckins  int
	UniqueTickets  int
	ByGate         map[string]int
	LastCheckinAt  *time.Time
}

File: frontend/src/lib/api/types/checkin.ts
// ── Domain types ──

export type CheckInStatus = "success" | "invalid" | "already_used" | "cancelled";

export interface CheckIn {
  id: string;
  ticket_id: string;
  event_id: string;
  user_id?: string;
  qr_code?: string;
  status: CheckInStatus;
  staff_id?: string;
  device_id?: string;
  gate?: string;
  notes?: string;
  check_in_time: string;
  created_at: string;
}

export interface EventCheckInStats {
  event_id: string;
  total_checkins: number;
  unique_tickets: number;
  by_gate: Record<string, number>;
  last_checkin_at?: string;
}

// ── Request types ──

export interface CheckInRequest {
  ticket_id: string;
  qr_code?: string;
  event_id: string;
  staff_id?: string;
  device_id?: string;
  gate?: string;
}
